include/seqan3/io/sequence/sequence_file_in.hpp

```diff
@@ -14,155 +58,123 @@ namespace seqan3
-    using sequence_file_in_traits::valid_formats;             // = std::variant<gzip_compressor, sequence_file_format_bam>
-
-    // constructor with arg
-    sequence_file_in(std::string const & _file_name);
-
-    // copy construction and assignment are deleted
-    // implicitly because we don't want multiple access to file
+    /* types */
+    using stream_type = typename detail::file_base<sequence_file_in_traits>::stream_type;
+    using valid_format_types = typename detail::file_base<sequence_file_in_traits>::valid_format_types;
+
+    /* constructors */
+    sequence_file_in(std::experimental::filesystem::path _file_name) :
+        detail::file_base<sequence_file_in_traits>(_file_name) {};
+    sequence_file_in() = delete;
     sequence_file_in(sequence_file_in const &) = delete;
     sequence_file_in & operator=(sequence_file_in const &) = delete;
-
-    // move construction and assignment are defaulted
     sequence_file_in(sequence_file_in &&) = default;
     sequence_file_in & operator=(sequence_file_in &&) = default;
-
-    //TODO make the requirements stricter
-    template <typename sequence_type, typename id_type, typename qual_type>
-        requires container_concept<sequence_type> &&
-                 container_concept<id_type> &&
-                 container_concept<qual_type>
-    void read(sequence_type && seq, id_type && id, qual_type && qual);
-
-    template <typename seqs_type, typename ids_type, typename quals_type, size_t max_records = 0>
-        requires container_concept<typename seqs_type::value> &&
-                 container_concept<typename ids_type::value> &&
-                 container_concept<typename quals_type::value>
-    void read(seqs_type && seqs, ids_type && ids, quals_type && quals);
+    ~sequence_file_in() = default;
 
     /* options */
     struct options_type
     {
         // post-processing filters that operate on buffer before assignment to out-value
-        std::function<void(std::string &)>  sequence_filter = [] (std::string & seq)  {};
-        std::function<void(std::string &)>   id_filter = [] (std::string & id)   {};
+        std::function<void(std::string &)> sequence_filter = [] (std::string & seq) {};
+        std::function<void(std::string &)> meta_filter = [] (std::string & meta) {};
         std::function<void(std::string &)> qual_filter = [] (std::string & qual) {};
     };
     options_type options;
 
-protected:
-    ~sequence_file_in() = default;
-
-private:
-    /* file format */
-    std::string file_name;
-    stream_type stream;
-    valid_formats format;
+    /* member functions */
+    // TODO make the requirements stricter
+    template <typename sequence_type, typename meta_type, typename qual_type>
+        requires sequence_concept<std::decay_t<sequence_type>> &&
+                 sequence_concept<std::decay_t<meta_type>> &&
+                 sequence_concept<std::decay_t<qual_type>>
+    inline void read(sequence_type && seq,
```

Comment:
you define but not declare the function ```inline```, according to https://isocpp.org/wiki/faq/inline-functions.